doc/sphinx/proof-engine/ltac.rst

```diff
@@ -1482,12 +1497,45 @@ expression returns an identifier:
    .. todo you can't have a :tacn: with the same name as a :gdef: for now,
       eg `fresh` can't be both
 
-   Returns a fresh identifier name (i.e. one that is not already used in the context).
+   Returns a fresh identifier name (i.e. one that is not already used in the context
+   and not previously returned by :tacn:`fresh` in the current :token:`ltac_expr`).
    The fresh identifier is formed by concatenating the final :token:`ident` of each :token:`qualid`
    (dropping any qualified components) and each specified :token:`string`.
    If the resulting name is already used, a number is appended to make it fresh.
    If no arguments are given, the name is a fresh derivative of the name ``H``.
 
+   .. note:: We recommend generating the fresh identifier immediately before
+      adding it in the proof context.  Using :tacn:`fresh` in a local function
+      may not work as you expect:
+
+      Successive :tacn:`fresh`\es give distinct names even if the names haven't
+      yet been added to the proof context:
+
+      .. coqtop:: reset none
+
+         Goal True -> True.
+
+      .. coqtop:: out
+
+         intro x.
+
+      .. coqtop:: all
+
+         let a := fresh "x" in
+         let b := fresh "x" in
+         idtac a b.
+
+      When applying :tacn:`fresh` in a function, the name is chosen based on the
+      proof context at the point where the function was defined:
```

Comment:
```suggestion
      tactic context at the point where the function was defined:
```
c.f.
```coq
Goal True -> True.
  let f := fun _ => fresh "H" in
  intro H;
    let n := f () in
    idtac n. (* H0, not H, even though H wasn't in the proof context *)
```

Reply:
Will change, although "tactic context" is not defined anywhere.